Replaying events

alexandergeegds opened the pull request:
Adds scripts to:

- download logs from CloudWatch
    - you have to authenticate with AWS first
- clean, filter and sort the download, ready for replaying
- prepare a Docker container ready to replay the data into Matomo
    - the script builds a Docker image, runs a container, and opens the command line; the command that needs to be run requires Matomo credentials, so is not currently automated.

There aren't any tests for anything, but I'm also not entirely sure what we can test. Maybe that the input and output formats of the filter script are what we expect, but is that worthwhile adding? Anything else?

alexandergeegds committed:
Add python script to filter logs from AWS

This assumes the logs have been downloaded via a call to the AWS CLI
via `aws logs filter-log-events`.

alexandergeegds committed:
Add some basic reporting to the filter

This allows the filter to report on how many messages were included
or excluded, which is useful for verifying the number of messages
against AWS CloudWatch. If a JSON decoding error is encountered,
the report also allows the user to determine how far through the
input file the tool got before the error was encountered.

alexandergeegds committed:
Ignore everything except the script

So that filtered files cannot be accidentally committed.

alexandergeegds committed:
Add scriptlet to download Matomo error logs

This script, when given a start and end time in milliseconds past the
unix epoch, downloads any Matomo Nginx logs that resulted in 5XX
responses. For example, this could be used to retrieve access logs
from a period of Matomo downtime.

alexandergeegds committed:
Add helpers for replaying events into Matomo

The provided Dockerfile uses a clone of a fork of the Matomo
analytics script, capable of handling `X-Forwarded-For` headers.
This, and a pre-defined access log are copied into the image.
A template command to replay events is echoed into the terminal -
a Matomo password currently has to be entered in plaintext, so should
not be baked into the Docker image.
The container is then spun up for the developer to enter the command.

This setup is clearly not ideal. But it is functional.
On the basis that something is better than nothing, it should be
iterated on as improvements become clearer.

alexandergeegds committed:
Add some brief explanations to the Readme

The Readme now provides a brief overview of the purpose of each
script, as well as basic instructions on running them.
As noted in the Readme, there is no context in these files.
This is a purposeful decision.

alexandergeegds committed:
Sort log messages by timestamp: earliest first

The Matomo log analytics tool states in its requirements that
"inside each log file the log lines must be ordered by date".
(https://matomo.org/docs/log-analytics-tool-how-to/)

Therefore, this script now ensures that the timestamp from the logs
are used to sort the messages before outting the transformed list.
This should make this script a sufficient step prior to replaying
logs downloaded from AWS CloudWatch into Matomo.

alexandergeegds committed:
Add a reminder to authenticate with AWS

chrisclayson commented on README.md:
```diff
@@ -1,2 +1,74 @@
+docker build -t matomo-replay-events .
+docker run --rm -it matomo-replay-events
+# inside docker container:
+python -u /log-analytics/import_logs.py \
```
Any reason you wouldn't make this command the entry point for Docker image and bind mount the access.log file at runtime?

alexandergeegds replied:
I didn't want to make it the entrypoint because of needing the credentials. You'd have to override all the arguments to specify the correct credentials, so I thought it was cleaner to just drop into a shell if you run the container without a specific override. Is there a better way of doing all this? (Short of getting the PHP auth to work... But the error messages I was getting were not particularly useful...)

As for the bind mount, you're right, that would be better. Especially for a large log file. I'll update. 👍

chrisclayson replied:
OK ... so, up to you, but you could define the credentials as environment variables, change the `ENTRYPOINT` in `Dockerfile` to `"python"`, and then change the `CMD` to:

```
CMD = [ 
   "-u",  
   "/log-analytics/import_logs.py",
    "--url=$MATAMO_URL" ,
    "--log-format-name=nginx_json",
    "--replay-tracking",
    "--enable-static",
    "--enable-bots",
    "--enable-reverse-dns",
    "--recorders=2",
    "--login=$MATOMO_LOGIN",
    "--password='$MATOMO_PASSWORD'",
    "/access.log"
]
```
N.B. Using `CMD` allows you to override the parameters on the dockker command line without changing the entrypont.

You can then run the image with:
```
docker run --rm \
    -v "$(pwd)"/access.log:/access.log \
    -e MATOMO_URL=<url> \
    -e MATOMO_LOGIN=<username> \
    -e MATOMO_PASSWORD=<password> \
    -it matomo-import
```

alexandergeegds replied:
Ahh, thanks. That's helpful. I didn't know about the `-e` flag. I'll make some changes based on this and update the Readme. I guess I should add checks for non-empty values for those environment variables into the bash script too.

alexandergeegds replied:
Right, I think I've found a combination that works. Thanks for all your ideas and help. Any further improvements that can be made, do suggest!

alexandergeegds committed:
Use a bind mount instead of copying

When dealing with large log files, copying the file into the Docker
image will create several redundant copies of the data. Using a bind
mount is better in almost every respect: data isn't stored in the
image, nor unnecessarily copied.

alexandergeegds committed:
Use a consistent docker image tag

The tag name in the helper script has been updated to match that
suggested by the Readme. The Readme itself has also been updated to
make use of the script.

alexandergeegds committed:
Authenticate for Matomo via environment variables

Instead of requiring a user manually run a command inside the
resultant Docker container, they now simply need to specify their
credentials via certain environment variables. Those values are
then passed into the container at run-time.

robinmitra commented on README.md:
```diff
@@ -1,2 +1,76 @@
+
+This script can be used to download logs from AWS CloudWatch.
+It assumes the logs in question are inside a log group called "matomo".
+Pass it two millisecond timestamps and it will download any errored requests from Nginx between those timestamps.
```
```suggestion
Pass it two millisecond-timestamps and it will download any errored requests from Nginx between those timestamps.
```

alexandergeegds committed:
Improve grammar

Co-Authored-By: Robin Mitra <robinmitra1@gmail.com>

robinmitra commented on replay-events/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM python:2-alpine
+
+COPY --from=analytics-fork /matomo-log-analytics /log-analytics
+COPY docker-entrypoint.sh /entrypoint.sh
```
Not completely obvious to me, what is the need for multi-stage build here?

alexandergeegds replied:
I think all it's doing is keeping git out of the resultant image. Probably overkill here, but originally I was doing more in other images than this. Can remove and just install git in the image if you don't think it's worth the added complexity?

robinmitra replied:
Probably not worth having a second build in there, but worth adding a one-liner to delete the `.git` directory.

robinmitra commented on download-logs/download.sh:
```diff
@@ -0,0 +1,22 @@
+        > aws_logs.txt
+    exit $?
+fi
+exit 1
```
```suggestion
```

Is this really an error/failure? It feels like the script is treating a user's decision to not proceed as a failure, which doesn't seem right. 

As a comparison, `rm -i` prompts user for confirmation to delete, and returns a status of `0` even if the user answers `n`. If the user terminates the command (`ctrl+c`), it treats it as a failure, which makes sense as the command didn't complete.

alexandergeegds replied:
That's a fair point. I based this off a macro I had already written for other things, where not proceeding was meant to fail. But it's not appropriate here.

If we remove this, can we also get rid of the `exit $?` above, do you know? I only added that to prevent failing on exiting the conditional.

alexandergeegds replied:
I've tried doing something similar in a small example, and it looks like that `exit $?` is still needed unless I `set -e` in the script. Any preference as to which approach we use?

robinmitra replied:
I'm leaning towards `set -e`.

robinmitra commented on replay-events/.gitignore:
```diff
@@ -0,0 +1,5 @@
+!.gitignore
+!docker-entrypoint.sh
+!Dockerfile
+!replay.sh
```
This feels a little backwards to me in that it will need to be updated every time a file needs to be added. Why not just ignore the files that actually need to be ignored (e.g. *.log)? It doesn't appear that the directory is going to have an arbitrary number of files with unpredictable names/extensions. There might be a good reason for this, so feel free to tell me if I'm missing the point.

This applies to other gitignore files in the PR as well.

alexandergeegds replied:
I do this precisely because I didn't want to make it possible to accidentally include log files with personal information in a git commit. (Which seems pertinent given what happened recently...) Since we don't actually want to track many files anyway, explicitly adding the few we do want to track here serves as a measure of protection, especially given the nature of the log files being generated.

If this repo were to start getting used more widely, or we were planning to expand it significantly, I'd agree it may not be helpful. But maybe we reevaluate at that point, rather than now?

robinmitra replied:
Happy with your reasoning 👍

alexandergeegds committed:
Remove multistage build

The multi-stage build only saves on an install of git. Including it
without other intermediate steps is probably excessive. Instead, the
.git folder within the repository clone is removed to save a little
space within that layer of the image.

alexandergeegds committed:
Improve error status responses when downloading

If a user has decided not to continue, then terminating the script
after having completing vacuously should not result in a non-zero
exit code. That is, completing should be successful.
Further, by setting the `-e` flag, there is no need to explicitly
exit with the code from the AWS CLI call - a non-zero exit code will
automatically cause the script to terminate and yield that code.

alexandergeegds committed:
Update documentation reference for access log

With recent changes to the Docker image, the file name can no longer
be changed in the Dockerfile. The file is made available through a
bind mount at container run-time, so it is there that the name needs
to change.

alexandergeegds commented:
Pushed some changes as per @robinmitra 's review. 🙇

robinmitra reviewed (dismissed):
👍

alexandergeegds committed:
Swap credentials with access token

alexandergeegds commented:
I have pushed a small change to use the Matomo authentication token rather than login credentials, as @adityapahuja demonstrated.

Given that there is now also https://github.com/alphagov/verify-matomo-tool we may want to consolidate these repos at some point. However, for now, I think it's still worth merging this so we at least have these scripts available, and look to merge as a future changeset.

robinmitra reviewed (approved):
👍
